From 3cc200ae134be963eac9b6e00e6b0f03e4f473ce Mon Sep 17 00:00:00 2001
From: Richard Gong <richard.gong@intel.com>
Date: Tue, 22 Jun 2021 12:24:41 -0500
Subject: [PATCH 09/42] firmware: stratix10-svc: update for getting the service
 key info

commit 34e4749e000f0fcdfdd1d02de58682abc48e0ae1 from
https://github.com/altera-opensource/linux-socfpga/commits/socfpga-5.4.124-lts

Update to support getting the service key info.

Signed-off-by: Richard Gong <richard.gong@intel.com>
Signed-off-by: Siew Chin Lim <elly.siew.chin.lim@intel.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 drivers/firmware/stratix10-svc.c              | 11 +++++++-
 include/linux/firmware/intel/stratix10-smc.h  | 27 +++++++++++++++++++
 .../firmware/intel/stratix10-svc-client.h     |  4 +++
 3 files changed, 41 insertions(+), 1 deletion(-)

diff --git a/drivers/firmware/stratix10-svc.c b/drivers/firmware/stratix10-svc.c
index 3effac4ab2b9..e6173dbae2a7 100644
--- a/drivers/firmware/stratix10-svc.c
+++ b/drivers/firmware/stratix10-svc.c
@@ -375,6 +375,7 @@ static void svc_thread_recv_status_ok(struct stratix10_svc_data *p_data,
 	case COMMAND_FCS_ATTESTATION_MEASUREMENTS:
 	case COMMAND_FCS_ATTESTATION_CERTIFICATE:
 	case COMMAND_FCS_CRYPTO_EXPORT_KEY:
+	case COMMAND_FCS_CRYPTO_GET_KEY_INFO:
 		cb_data->status = BIT(SVC_STATUS_OK);
 		cb_data->kaddr2 = svc_pa_to_va(res.a2);
 		cb_data->kaddr3 = &res.a3;
@@ -605,12 +606,12 @@ static int svc_normal_to_secure_thread(void *data)
 			a2 = 0;
 			break;
 
+		/* for service key management */
 		case COMMAND_FCS_CRYPTO_IMPORT_KEY:
 			a0 = INTEL_SIP_SMC_FCS_IMPORT_CRYPTO_SERVICE_KEY;
 			a1 = (unsigned long)pdata->paddr;
 			a2 = (unsigned long)pdata->size;
 			break;
-
 		case COMMAND_FCS_CRYPTO_EXPORT_KEY:
 			a0 = INTEL_SIP_SMC_FCS_EXPORT_CRYPTO_SERVICE_KEY;
 			a1 = pdata->arg[0];
@@ -623,6 +624,13 @@ static int svc_normal_to_secure_thread(void *data)
 			a1 = pdata->arg[0];
 			a2 = pdata->arg[1];
 			break;
+		case COMMAND_FCS_CRYPTO_GET_KEY_INFO:
+			a0 = INTEL_SIP_SMC_FCS_GET_CRYPTO_SERVICE_KEY_INFO;
+			a1 = pdata->arg[0];
+			a2 = pdata->arg[1];
+			a3 = (unsigned long)pdata->paddr_output;
+                        a4 = (unsigned long)pdata->size_output;
+			break;
 
 		/* for polling */
 		case COMMAND_POLL_SERVICE_STATUS:
@@ -714,6 +722,7 @@ static int svc_normal_to_secure_thread(void *data)
 			case COMMAND_FCS_CRYPTO_IMPORT_KEY:
 			case COMMAND_FCS_CRYPTO_EXPORT_KEY:
 			case COMMAND_FCS_CRYPTO_REMOVE_KEY:
+			case COMMAND_FCS_CRYPTO_GET_KEY_INFO:
 				cbdata->status = BIT(SVC_STATUS_INVALID_PARAM);
 				cbdata->kaddr1 = NULL;
 				cbdata->kaddr2 = NULL;
diff --git a/include/linux/firmware/intel/stratix10-smc.h b/include/linux/firmware/intel/stratix10-smc.h
index ada0cf6ed027..43eec1983b47 100644
--- a/include/linux/firmware/intel/stratix10-smc.h
+++ b/include/linux/firmware/intel/stratix10-smc.h
@@ -896,4 +896,31 @@ INTEL_SIP_SMC_FAST_CALL_VAL(INTEL_SIP_SMC_FUNCID_FPGA_CONFIG_COMPLETED_WRITE)
 #define INTEL_SIP_SMC_FCS_REMOVE_CRYPTO_SERVICE_KEY \
 	INTEL_SIP_SMC_FAST_CALL_VAL(INTEL_SIP_SMC_FUNCID_FCS_REMOVE_CRYPTO_SERVICE_KEY)
 
+/**
+ * Request INTEL_SIP_SMC_FCS_GET_CRYPTO_SERVICE_KEY_INFO
+ * Sync call to query the crypto service keys on the device
+ *
+ * Call register usage:
+ * a0 INTEL_SIP_SMC_FCS_GET_CRYPTO_SERVICE_KEY_INFO
+ * a1 session ID
+ * a2 key UID
+ * a3 physical address of the reponse data
+ * a4 max size of the response data (36 words with header)
+ * a3-a7 not used
+ *
+ * Return status:
+ * a0 INTEL_SIP_SMC_STATUS_OK, INTEL_SIP_SMC_STATUS_NOT_SUPPORTED or
+ *    INTEL_SIP_SMC_STATUS_ERROR
+ * a1 mailbox and status errors if a0 is INTEL_SIP_SMC_STATUS_ERROR
+ * 	31:24 -- reserved
+ * 	23:16 -- import/export/removal status error
+ * 	15:11 -- reserved
+ * 	10:0  -- mailbox error
+ * a2 physical address of the reponse data
+ * a3 size of the response data
+ */
+#define INTEL_SIP_SMC_FUNCID_FCS_GET_CRYPTO_SERVICE_KEY_INFO 115
+#define INTEL_SIP_SMC_FCS_GET_CRYPTO_SERVICE_KEY_INFO \
+	INTEL_SIP_SMC_FAST_CALL_VAL(INTEL_SIP_SMC_FUNCID_FCS_GET_CRYPTO_SERVICE_KEY_INFO)
+
 #endif
diff --git a/include/linux/firmware/intel/stratix10-svc-client.h b/include/linux/firmware/intel/stratix10-svc-client.h
index f456c0211fe7..9108a59b658b 100644
--- a/include/linux/firmware/intel/stratix10-svc-client.h
+++ b/include/linux/firmware/intel/stratix10-svc-client.h
@@ -183,6 +183,9 @@ struct stratix10_svc_chan;
  *
  * @COMMAND_FCS_CRYPTO_REMOVE_KEY: remove the crypto service key object
  * from the device, return status is SVC_STATUS_OK or SVC_STATUS_ERROR
+ *
+ * @COMMAND_FCS_CRYPTO_GET_KEY_INFO: get the crypto service key object
+ * info, return status is SVC_STATUS_OK or SVC_STATUS_ERROR
  */
 enum stratix10_svc_command_code {
 	/* for FPGA */
@@ -227,6 +230,7 @@ enum stratix10_svc_command_code {
 	COMMAND_FCS_CRYPTO_IMPORT_KEY,
 	COMMAND_FCS_CRYPTO_EXPORT_KEY,
 	COMMAND_FCS_CRYPTO_REMOVE_KEY,
+	COMMAND_FCS_CRYPTO_GET_KEY_INFO,
 };
 
 /**
-- 
2.31.1

